### PR TITLE
[RLlib] Rewrite PPO to use training_iteration + enable DD-PPO for Win32.

### PR DESCRIPTION
**Pull request by smorad**
## Why are these changes needed?

I'm trying to benchmark pytorch PPO performance, but this is difficult with the `execution_plan` api and associated DataFlow programming paradigm. This PR enables the use of the new experimental imperative `training_iteration` API. This also significantly aids in readability, and appears to provide a marginal (~7%) speedup.

Orange is before fix and pink is after fix, using config
```yaml
cartpole-ppo:
    env: CartPole-v0
    run: PPO
    stop:
        timesteps_total: 300000
    config:
        min_time_s_per_reporting: 5
        _disable_execution_plan_api: true,
        # Works for both torch and tf.
        framework: torch
        gamma: 0.99
        lr: 0.0003
        num_workers: 0
        observation_filter: MeanStdFilter
        num_sgd_iter: 6
        vf_loss_coeff: 0.01
        model:
            fcnet_hiddens: [32]
            fcnet_activation: linear
            vf_share_layers: true
```
Screen Shot 2022-04-02 at 9 39 18 PM

Another benchmark on Atari (Pong), using 15 workers: This indicates that w/o the changes, PPO has a much harder time to learn, but I'll need to confirm this more properly (more seeds):

Config:
```
# On a single GPU, this achieves maximum reward in ~15-20 minutes.
#
# $ python train.py -f tuned_configs/pong-ppo.yaml
#
pong-ppo:
    env: PongNoFrameskip-v4
    stop:
        episode_reward_mean: 19.0
    run: PPO
    config:
        # Works for both torch and tf.
        framework: tf
        lambda: 0.95
        kl_coeff: 0.5
        clip_rewards: True
        clip_param: 0.1
        vf_clip_param: 10.0
        entropy_coeff: 0.01
        train_batch_size: 5000
        rollout_fragment_length: 20
        sgd_minibatch_size: 500
        num_sgd_iter: 10
        num_workers: 15
        num_envs_per_worker: 5
        batch_mode: truncate_episodes
        observation_filter: NoFilter
        num_gpus: 1
        model:
            dim: 42
            vf_share_layers: true

        _disable_execution_plan_api:
            grid_search: [true, false]
```

Results:
```
== Status ==
Current time: 2022-04-08 04:36:59 (running for 00:22:00.63)
Memory usage on this node: 27.1/239.9 GiB
Using FIFO scheduling algorithm.
Resources requested: 32.0/32 CPUs, 2.0/4 GPUs, 0.0/147.65 GiB heap, 0.0/67.27 GiB objects
Result logdir: /home/ray/ray_results/pong-ppo
Number of trials: 2/2 (2 RUNNING)
+------------------------------------+----------+-------------------+-------------------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------+
| Trial name                         | status   | loc               | _disable_execution_plan_api   |   iter |   total time (s) |      ts |   reward |   episode_reward_max |   episode_reward_min |   episode_len_mean |
|------------------------------------+----------+-------------------+-------------------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------|
| PPO_PongNoFrameskip-v4_203c6_00000 | RUNNING  | 10.0.67.156:17005 | True                          |    327 |          1280.29 | 3237300 |    19.02 |                   21 |                   12 |            7343.62 |
| PPO_PongNoFrameskip-v4_203c6_00001 | RUNNING  | 10.0.67.156:17006 | False                         |    638 |          2522.73 | 6316200 |    19.15 |                   21 |                   -3 |            8530.15 |
+------------------------------------+----------+-------------------+-------------------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------+
```

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by gjoliver** [comment]
what's the relationship between this and https://github.com/ray-project/ray/pull/23686 ??
which one should we reivew?

**Comment by smorad** [comment]
> what's the relationship between this and #23686 ?? which one should we reivew?

Doh, I did not see this open PR. This implements identical capabilities to #23686. @sven1977 it's probably worth double checking against mine to see if there are any important differences. I think the main difference is this PR still warns about bad weight/clipping scales like the original verison.

**Comment by sven1977** [comment]
Oh no! :D Sorry @smorad , I didn't see your PR before I did something similar to this yesterday.
But no problem. Yours actually looks like it solved some of the problems my PR still had (e.g. PPO reward range checking).
Let me take a look ...

File: rllib/agents/ppo/ppo.py [hunk 1]
```diff
@@ -462,3 +477,54 @@ def execution_plan(
...
+            ):
+                logger.warning(
+                    "The magnitude of your value function loss for policy: {} is "
+                    "extremely large ({}) compared to the policy loss ({}). This "
+                    "can prevent the policy from learning. Consider scaling down "
+                    "the VF loss by reducing vf_loss_coeff, or disabling "
+                    "vf_share_layers.".format(policy_id, scaled_vf_loss, policy_loss)
+                )
+            # Warn about bad clipping configs
+            mean_reward = rollouts["rewards"][rollouts["agent_index"] == i].mean()
```
Comment (sven1977) [inline rllib/agents/ppo/ppo.py:522]: Great idea just doing this here and in this fashion! True, we don't really need entire episodes for this avg-reward estimate.

File: rllib/execution/rollout_ops.py [hunk 2]
```diff
@@ -337,6 +337,35 @@ def __call__(self, samples: SampleBatchType) -> SampleBatchType:
         return samples
 
 
+def standardize_fields(samples: SampleBatchType, fields: List[str]) -> SampleBatchType:
```
Comment (sven1977) [inline rllib/execution/rollout_ops.py:340]: Actually, we don't need this, I think. Can we just move the following code snippet directly into the iteration fn?

```
        # Standardize `advantages` values in train_batch.
        for policy_id, batch in train_batch.policy_batches.items():
            if Postprocessing.ADVANTAGES not in batch:
                raise KeyError(
                    f"`{Postprocessing.ADVANTAGES}` not found in SampleBatch for "
                    f"policy `{policy_id}`! Maybe this policy fails to add "
                    f"`{Postprocessing.ADVANTAGES}` in its `postprocess_trajectory` "
                    f"method? Or this policy is not meant to learn at all and you "
                    "forgot to remove it from the list under `config."
                    "multiagent.policies_to_train`."
                )
            batch[Postprocessing.ADVANTAGES] = standardized(
                batch[Postprocessing.ADVANTAGES])
```

File: rllib/execution/rollout_ops.py [hunk 3]
```diff
@@ -337,6 +337,35 @@ def __call__(self, samples: SampleBatchType) -> SampleBatchType:
...
+    """Standardize fields of the given SampleBatch"""
```
Comment (sven1977) [inline rllib/execution/rollout_ops.py:341]: Actually, could we just simplify this even further and directly add the following code to the `training_iteration()` method:

```
        # Standardize `advantages` values in train_batch.
        for policy_id, batch in train_batch.policy_batches.items():
            if Postprocessing.ADVANTAGES not in batch:
                raise KeyError(
                    f"`{Postprocessing.ADVANTAGES}` not found in SampleBatch for "
                    f"policy `{policy_id}`! Maybe this policy fails to add "
                    f"`{Postprocessing.ADVANTAGES}` in its `postprocess_trajectory` "
                    f"method? Or this policy is not meant to learn at all and you "
                    "forgot to remove it from the list under `config."
                    "multiagent.policies_to_train`."
                )
            batch[Postprocessing.ADVANTAGES] = standardized(
                batch[Postprocessing.ADVANTAGES])

```

File: rllib/agents/ppo/ppo_torch_policy.py [hunk 4]
```diff
@@ -152,6 +152,7 @@ def reduce_mean_valid(t):
             mean_vf_loss = reduce_mean_valid(vf_loss_clipped)
         # Ignore the value function.
         else:
+            value_fn_out = 0
```
Comment (sven1977) [inline rllib/agents/ppo/ppo_torch_policy.py:155]: Great fix! I guess this is already merged from the other PR.

File: rllib/agents/ppo/ppo.py [hunk 5]
```diff
@@ -462,3 +477,54 @@ def execution_plan(
...
             lambda result: warn_about_bad_reward_scales(config, result)
         )
+
+    @ExperimentalAPI
+    def training_iteration(self) -> ResultDict:
+        # Collect SampleBatches from sample workers
+        rollouts = synchronous_parallel_sample(self.workers)
+
+        # Concatenate the SampleBatches from each worker into one large SampleBatch
+        rollouts = SampleBatch.concat_samples(rollouts)
```
Comment (sven1977) [inline rllib/agents/ppo/ppo.py:487]: We might need some `train_batch_size` check here (while loop just like in rllib/agents/pg/pg.py).
Otherwise, the train batch size might be too small after just a single round of parallel `RolloutWorker.sample()` collection.

**Review by sven1977 — COMMENTED** [review]
Looks really great. Just a few nits.

**Comment by sven1977** [comment]
@gjoliver , @smorad , we'll move forward with this PR. Please ignore https://github.com/ray-project/ray/pull/23686
This PR solves the reward checking problem. The other PR (https://github.com/ray-project/ray/pull/23686) solves the batch size problem, the solution of which we should move into here.

File: rllib/agents/ppo/ppo.py (hunk 5 above)
Comment (gjoliver) [reply rllib/agents/ppo/ppo.py:487]: make it into a util to collect exactly train_batch # of samples from available workers in truncated_episode mode, and at least train_batch # of samples in complete_episode mode?

File: rllib/agents/ppo/ppo.py (hunk 1 above)
Comment (gjoliver) [reply rllib/agents/ppo/ppo.py:522]: should we log 1 in N times or something? won't this flood the training logs?

File: rllib/execution/rollout_ops.py (hunk 3 above)
Comment (smorad) [reply rllib/execution/rollout_ops.py:341]: This would require two identical for loops, which are quite ugly IMO. This is because we need to train AFTER normalizing, not before.

**Comment by smorad** [comment]
I've implemented nearly all suggestions and the code trains. I have not written unit tests and this is quite an important piece of code to be tested IMO. Unfortunately, I'm leaving on holiday very soon! We can either merge now and I will write tests when I return, or we can wait until I return to add tests to this PR.

**Review by gjoliver — APPROVED** [review]
nice! I vote we merge this as long as you promise to write the tests after you come back :)

File: rllib/execution/rollout_ops.py (hunk 3 above)
Comment (sven1977) [reply rllib/execution/rollout_ops.py:383]: Ok, fair enough. Fine with me :)

**Review by sven1977 — APPROVED** [review]
LGTM. Thanks for this PR @smorad !

**Comment by sven1977** [comment]
Fixed LINTer, just waiting for tests to pass.

**Comment by sven1977** [comment]
Fixed an older bug in multi_gpu_train_one_step (sgd_num_iter instead of num_sgd_iter :( ). This was causing PPO not to learn. It's fixed now and I ran several tests on Pendulum and CartPole. I'm not too concerned, this broke other benchmarks, as PPO is fairly sequential in its execution pattern. Will merge as soon as all tests pass again ...

Also, switched training_iteration to True by default.

File: rllib/agents/ppo/ppo.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo.py]: I think @smorad fixed this by a `log_once` if-block, which seems like always the best option for these warnings.

File: rllib/agents/ppo/ppo.py [hunk 6]
```diff
@@ -95,6 +109,7 @@ def __init__(self):
         self.train_batch_size = 4000
         self.lr = 5e-5
         self.model["vf_share_layers"] = False
+        self._disable_execution_plan_api = True
```
Comment (ArturNiederfahrenhorst) [inline rllib/agents/ppo/ppo.py:112]: Do we need this?

File: rllib/agents/ppo/ppo.py [hunk 7]
```diff
@@ -410,6 +426,83 @@ def get_default_policy_class(self, config: TrainerConfigDict) -> Type[Policy]:
         else:
             return PPOTFPolicy
 
+    @ExperimentalAPI
+    def training_iteration(self) -> ResultDict:
+        # Collect SampleBatches from sample workers until we have a full batch.
+        if self._by_agent_steps:
+            train_batch = synchronous_parallel_sample(
+                worker_set=self.workers, max_agent_steps=self.config["train_batch_size"]
```
Comment (ArturNiederfahrenhorst) [inline rllib/agents/ppo/ppo.py:434]: max_env_steps and max_env_steps should be used more often in other training iteration functions where we always wait until we have _at least_ enough samples to continue and therefore sometimes don't use exact batch or samples sizes!

File: rllib/agents/trainer.py [hunk 8]
```diff
@@ -1133,6 +1133,23 @@ def step(self) -> ResultDict:
...
+                )
+
+            def fn(env, env_context, task_fn):
+                new_task = task_fn(result, env, env_context)
+                cur_task = env.get_task()
+                if cur_task != new_task:
+                    env.set_task(new_task)
+
+            fn = functools.partial(fn, task_fn=self.config["env_task_fn"])
+            self.workers.foreach_env_with_context(fn)
```
Comment (ArturNiederfahrenhorst) [inline rllib/agents/trainer.py:1151]: Can this time out and create errors that we don't recover from properly due to this being outside of step_attempt()?

File: rllib/agents/trainer.py [hunk 9]
```diff
@@ -1432,30 +1432,18 @@ def training_iteration(self) -> ResultDict:
...
-            new_sample_batches = synchronous_parallel_sample(self.workers)
-            sample_batches.extend(new_sample_batches)
-            num_env_steps += sum(len(s) for s in new_sample_batches)
-            num_agent_steps += sum(
-                len(s) if isinstance(s, SampleBatch) else s.agent_steps()
-                for s in new_sample_batches
+        # Collect SampleBatches from sample workers until we have a full batch.
+        if self._by_agent_steps:
+            train_batch = synchronous_parallel_sample(
+                worker_set=self.workers, max_agent_steps=self.config["train_batch_size"]
```
Comment (ArturNiederfahrenhorst) [inline rllib/agents/trainer.py:1438]: So I guess we are rolling out the max_agent_steps/max_env_steps thing with this PR? (Related comment above)

File: rllib/execution/train_ops.py [hunk 10]
```diff
@@ -147,40 +179,10 @@ def multi_gpu_train_one_step(trainer, train_batch) -> Dict:
     trainer._counters[NUM_ENV_STEPS_TRAINED] += train_batch.count
     trainer._counters[NUM_AGENT_STEPS_TRAINED] += train_batch.agent_steps()
 
-    # Update weights - after learning on the local worker - on all remote
```
Comment (ArturNiederfahrenhorst) [inline rllib/execution/train_ops.py:150]: Although I approve of this, this might break other training iteration functions currently in progress I think.

File: rllib/execution/train_ops.py [hunk 11]
```diff
@@ -147,40 +179,10 @@ def multi_gpu_train_one_step(trainer, train_batch) -> Dict:
...
-        >>> rollouts = ParallelRollouts(...) # doctest: +SKIP
-        >>> train_op = rollouts.for_each(TrainOneStep(workers)) # doctest: +SKIP
-        >>> # This trains the policy on one batch.
-        >>> print(next(train_op)) # doctest: +SKIP
-        SampleBatch(...), {"learner_stats": ...}
-
-    Updates the STEPS_TRAINED_COUNTER counter and LEARNER_INFO field in the
-    local iterator context.
-    """
-
```
Comment (ArturNiederfahrenhorst) [inline rllib/execution/train_ops.py:183]: Just our of curiosity, why are we removing these docstrings?

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
I love the max_****_steps change, we should totally do this in other training iteration functions. Other than that I have some minor questions, nothing pressing :)
Sorry I'm late to the party 😄

File: rllib/execution/train_ops.py (hunk 11 above)
Comment (sven1977) [reply rllib/execution/train_ops.py:183]: Ah, sorry, b/c we are starting to deprecate things and very soon would like users to only use the `synchronous_parallel_sample` and `asynchronous_parallel_requests` functions, instead.

File: rllib/agents/ppo/ppo.py (hunk 6 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo.py:112]: Yes, the default is still `False`, so PPOConfig needs to set this here.

File: rllib/agents/ppo/ppo.py (hunk 7 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo.py:434]: Yes, totally agree, we should now unify this and remove the `while` loops everywhere else. We'll clean this up now that we are getting very close to setting `_disable_execution_plan_api=True` by default.

File: rllib/execution/train_ops.py (hunk 10 above)
Comment (sven1977) [reply rllib/execution/train_ops.py:150]: Sorry, yes, we need to be careful with this. Moved this out, b/c we should separate concerns: train and update should be separate steps handled by separate functions.
